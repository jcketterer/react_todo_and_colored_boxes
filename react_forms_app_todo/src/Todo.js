import React, { useState } from 'react'

const Todo = ({ task = 'Default Todo', id = 1, remove, update }) => {
  const [editTodo, setEditTodo] = useState(task)
  const [isEditing, setIsEditing] = useState(false)

  const switchEdit = () => {
    setIsEditing(edit => !edit)
  }

  const handleChange = e => {
    setEditTodo(e.target.value)
  }

  const handleDelete = () => remove(id)

  const handleUpdate = e => {
    e.preventDefault()
    update(id, editTodo)
    setIsEditing(false)
  }

  let JSX = (
    <div>
      <li>{task}</li>
      <button onClick={switchEdit}>Edit</button>
      <button onClick={handleDelete}>Remove</button>
    </div>
  )

  if (isEditing) {
    JSX = (
      <div>
        <form onSubmit={handleUpdate}>
          <input type="text" value={editTodo} onChange={handleChange} />
          <button>Update!</button>
        </form>
      </div>
    )
  }
  return JSX
}

export default Todo
