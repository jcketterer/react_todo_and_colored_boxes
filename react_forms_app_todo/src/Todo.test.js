import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import Todo from './Todo'

it('should render without breaking', () => {
  render(<Todo />)
})

it('matches snap', () => {
  const { asFragment } = render(<Todo />)
  expect(asFragment()).toMatchSnapshot()
})

it('should match snapshot when edited', () => {
  const { asFragment, getByText } = render(<Todo />)
  const editButton = getByText('Edit')
  fireEvent.click(editButton)
  expect(asFragment()).toMatchSnapshot()
})

it('should run the udpate function of form submit', () => {
  const update = jest.fn()
  const { getByText } = render(<Todo update={update} />)
  const edit = getByText('Edit')
  fireEvent.click(edit)
  const updateBtn = getByText('Update!')
  fireEvent.click(updateBtn)
  expect(update).toHaveBeenCalled()
})

it('should run the delete function on delete button click', () => {
  const remove = jest.fn()
  const { getByText } = render(<Todo remove={remove} />)
  const deleteBtn = getByText('Remove')
  fireEvent.click(deleteBtn)
  expect(remove).toHaveBeenCalled()
})
