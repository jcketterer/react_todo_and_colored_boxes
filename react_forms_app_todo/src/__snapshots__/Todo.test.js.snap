// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snap 1`] = `
<DocumentFragment>
  <div>
    <li>
      Default Todo
    </li>
    <button>
      Edit
    </button>
    <button>
      Remove
    </button>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot when edited 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        type="text"
        value="Default Todo"
      />
      <button>
        Update!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
